binary_search_tree/tree.py

```diff
@@ -9,51 +9,119 @@ def __init__(self, key, val = None):
         self.right = None
         
 
-
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def add_helper_function(self, current, next):
+        if next.key < current.key:
+            if not current.left:
+                current.left = next
+                return
+            self.add_helper_function(current.left, next)
+        else:
+            if not current.right:
+                current.right = next
+                return
+            self.add_helper_function(current.right, next)
+
     def add(self, key, value = None):
-        pass
+        if not self.root:
+            self.root = TreeNode(key, value)
+        else:
+            new_node = TreeNode(key, value)
+            self.add_helper_function(self.root, new_node)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
 
-    # Time Complexity: 
-    # Space Complexity: 
+ 
     def find(self, key):
-        pass
+        current = self.root
+
+        while current:
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                current = current.left
+            else:
+                current = current.right
+
+        return None
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
```

Comment:
Good news! Since you did an iterative implementation of `find`, you don't take the hit of the extra stack frames you get from recursion, so your space complexity here is actually O(1), since no new memory is allocated.